### ISSUE TITLE
High CPU Usage due to proxy.useCustomHeaders(req) in ds_proxy.go

### ISSUE THREAD

**Issue by m-a-leclercq** [grafana/grafana#20076]
**What happened**:
``` perf top -p <Grafana PID>```  shows high CPU usage for function ```sha256.block```

Below is a [flamegraph](https://github.com/brendangregg/FlameGraph) of CPU Usage:
![image](https://user-images.githubusercontent.com/13354100/67754359-9c5c8d80-fa36-11e9-9e46-36bfb4a838c6.png)

We found out that by commenting line 168 in [```pkg/api/pluginproxy/ds_proxy.go```](https://github.com/grafana/grafana/blob/v6.1.6/pkg/api/pluginproxy/ds_proxy.go#L168) the problem disappeared. This has not had any impact on our end users as far as we can tell and everything is still working as expected for our use case.

**How to reproduce it (as minimally and precisely as possible)**: 
Good question... We use provisioning with over 150 Elasticsearch Datasources all using Basic auth, TLS client cert and CA in Server Access mode. These three settings (``` tlsCACert, tlsClientCert, tlsClientKey```) are the only ones used in SecureJsonData.  We don't rewrite headers as far as I know.

We do have some auth.proxy settings in grafana.ini but it shouldn't impact that part of the code. 

**Anything else we need to know?**:
All of our Datasources are behind a TLS 2-Way check, and certs are still being checked properly by our reverse proxy. 

[Related post I made yesterday in the community](https://community.grafana.com/t/proxy-usecustomheaders-causing-high-cpu-load/22319)

**Environment**:
- Grafana version: 6.1.6
- Data source type & version: Elasticsearch 6.X
- OS Grafana is installed on: RHEL 7
- User OS & Browser: Chrome
- Grafana plugins:
- Others:

**Comment by aocenas** [comment]
@marefr Can you or somebody from backend team take a look at this? Seems like the code was move in https://github.com/grafana/grafana/pull/19508 but in general seems like we still do the same amount of work there. Also in the community somebody mentioned having the same issue in 6.3 https://community.grafana.com/t/proxy-usecustomheaders-causing-high-cpu-load/22319.

**Comment by m-a-leclercq** [comment]
Hi @aocenas, as mentioned in my post, I am the one who opened the topic you mention on the community. 

I am no Go expert, but it seems the ```if```condition to call the function is a bit too wide  ([from the official documentation](https://grafana.com/docs/administration/provisioning/#secure-json-data)) : 
```go
		// Lookup and use custom headers
		if proxy.ds.SecureJsonData != nil {
			proxy.useCustomHeaders(req)
		}
```
Don't know if the current codebase allows to check for```httpHeaderValue<int>``` inside SecureJsonData but maybe it should.

```pkg/components/securejsondata/securejsondata.go``` seems to be where that check should be done.

**Comment by m-a-leclercq** [comment]
This issue is potentially fixed by #19508 which reworked the way datasources attributes are accessed.

However, I have not yet tested it to confirm it does indeed definetely fix this problem.

Should I leave this issue open? 

Thanks.

**Comment by aknuds1** [comment]
I can look into this.

**Comment by marefr** [comment]
CustomHeaders are called once per datasource and cached since these changes https://github.com/grafana/grafana/pull/19508/files#diff-4c11e19862ae21a32030cc55dee48cb3R75 so above issue should be fixed in 6.5.

**Comment by marefr** [comment]
We had a similar problem #20579 which we released a fix for in 6.3.7, 6.4.5 and 6.5.0.

**Comment by m-a-leclercq** [comment]
Thank you for your reply @marefr. Closing this issue then 😄
